samza-core/src/main/java/org/apache/samza/coordinator/JobModelCalculator.java

```diff
@@ -0,0 +1,239 @@
+
+    // If no mappings are present (first time the job is running) we return -1, this will allow 0 to be the first change
+    // mapping.
+    int maxChangelogPartitionId = changeLogPartitionMapping.values().stream().max(Comparator.naturalOrder()).orElse(-1);
+    // Sort the groups prior to assigning the changelog mapping so that the mapping is reproducible and intuitive
+    TreeMap<TaskName, Set<SystemStreamPartition>> sortedGroups = new TreeMap<>(groups);
+    Set<TaskModel> taskModels = new HashSet<>();
+    for (Map.Entry<TaskName, Set<SystemStreamPartition>> group : sortedGroups.entrySet()) {
+      TaskName taskName = group.getKey();
+      Set<SystemStreamPartition> systemStreamPartitions = group.getValue();
+      Optional<Integer> changelogPartitionId = Optional.ofNullable(changeLogPartitionMapping.get(taskName));
+      Partition changelogPartition;
+      if (changelogPartitionId.isPresent()) {
+        changelogPartition = new Partition(changelogPartitionId.get());
+      } else {
+        // If we've never seen this TaskName before, then assign it a new changelog partition.
+        maxChangelogPartitionId++;
+        LOG.info(
+            String.format("New task %s is being assigned changelog partition %s.", taskName, maxChangelogPartitionId));
+        changelogPartition = new Partition(maxChangelogPartitionId);
+      }
+      taskModels.add(new TaskModel(taskName, systemStreamPartitions, changelogPartition));
+    }
+
+    // Here is where we should put in a pluggable option for the SSPTaskNameGrouper for locality, load-balancing, etc.
+    TaskNameGrouperFactory containerGrouperFactory =
+        ReflectionUtil.getObj(taskConfig.getTaskNameGrouperFactory(), TaskNameGrouperFactory.class);
+    boolean standbyTasksEnabled = jobConfig.getStandbyTasksEnabled();
+    int standbyTaskReplicationFactor = jobConfig.getStandbyTaskReplicationFactor();
+    TaskNameGrouperProxy taskNameGrouperProxy =
+        new TaskNameGrouperProxy(containerGrouperFactory.build(config), standbyTasksEnabled,
+            standbyTaskReplicationFactor);
+    Set<ContainerModel> containerModels;
+    boolean isHostAffinityEnabled = new ClusterManagerConfig(config).getHostAffinityEnabled();
+    if (isHostAffinityEnabled) {
+      containerModels = taskNameGrouperProxy.group(taskModels, grouperMetadata);
+    } else {
+      containerModels =
+          taskNameGrouperProxy.group(taskModels, new ArrayList<>(grouperMetadata.getProcessorLocality().keySet()));
+    }
+
+    Map<String, ContainerModel> containerMap =
+        containerModels.stream().collect(Collectors.toMap(ContainerModel::getId, Function.identity()));
+    return new JobModel(config, containerMap);
+  }
+
+  /**
+   * Refresh Kafka topic list used as input streams if enabled {@link org.apache.samza.config.RegExTopicGenerator}
+   * @param config Samza job config
+   * @return refreshed config
+   */
+  private static Config refreshConfigByRegexTopicRewriter(Config config) {
```

Comment:
Just to be consistent with the above, can we rename the parameter to `originalConfig`?